src/main/java/com/appdynamics/extensions/webspheremq/metricscollector/ReadConfigurationEventQueueCollector.java

```diff
@@ -0,0 +1,175 @@
+import com.appdynamics.extensions.logging.ExtensionsLoggerFactory;
+import com.appdynamics.extensions.metrics.Metric;
+import com.appdynamics.extensions.webspheremq.common.WMQUtil;
+import com.appdynamics.extensions.webspheremq.config.QueueManager;
+import com.appdynamics.extensions.webspheremq.config.WMQMetricOverride;
+import com.google.common.collect.Lists;
+import com.ibm.mq.MQException;
+import com.ibm.mq.MQGetMessageOptions;
+import com.ibm.mq.MQMessage;
+import com.ibm.mq.MQQueue;
+import com.ibm.mq.MQQueueManager;
+import com.ibm.mq.constants.CMQC;
+import com.ibm.mq.constants.CMQCFC;
+import com.ibm.mq.constants.MQConstants;
+import com.ibm.mq.headers.pcf.MQCFH;
+import com.ibm.mq.headers.pcf.PCFMessage;
+import com.ibm.mq.headers.pcf.PCFMessageAgent;
+import com.singularity.ee.agent.systemagent.api.exception.TaskExecutionException;
+import org.slf4j.Logger;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+
+public class ReadConfigurationEventQueueCollector implements MetricsPublisher {
+
+	private static final Logger logger = ExtensionsLoggerFactory.getLogger(ReadConfigurationEventQueueCollector.class);
+	private final CountDownLatch countDownLatch;
+	private final MetricWriteHelper metricWriteHelper;
+	private final QueueManager queueManager;
+	private final PCFMessageAgent agent;
+	private final MonitorContextConfiguration monitorContextConfig;
+	private final MQQueueManager mqQueueManager;
+	private final Map<String, WMQMetricOverride> metricsToReport;
+	private final long bootTime;
+
+	public ReadConfigurationEventQueueCollector(Map<String, WMQMetricOverride> metricsToReport, MonitorContextConfiguration monitorContextConfig, PCFMessageAgent agent, MQQueueManager mqQueueManager, QueueManager queueManager, MetricWriteHelper metricWriteHelper, CountDownLatch countDownLatch) {
+		this.metricsToReport = metricsToReport;
+		this.monitorContextConfig = monitorContextConfig;
+		this.agent = agent;
+		this.mqQueueManager = mqQueueManager;
+		this.queueManager = queueManager;
+		this.metricWriteHelper = metricWriteHelper;
+		this.countDownLatch = countDownLatch;
+		this.bootTime = System.currentTimeMillis();
+	}
+
+	private PCFMessage findLastUpdate(long entryTime, String configurationQueueName) throws Exception {
+		// find the last update:
+		PCFMessage candidate = null;
+
+		boolean consumeEvents = this.queueManager.getConsumeConfigurationEventInterval() > 0 && (entryTime-this.bootTime) % this.queueManager.getConsumeConfigurationEventInterval() == 0;
+
+		MQQueue queue = null;
+		try {
+			int queueAccessOptions = MQConstants.MQOO_FAIL_IF_QUIESCING | MQConstants.MQOO_INPUT_SHARED;
+			if (!consumeEvents) {
+				// we are not consuming the events.
+				queueAccessOptions |= MQConstants.MQOO_BROWSE;
+			}
+			queue = mqQueueManager.accessQueue(configurationQueueName, queueAccessOptions);
+			int maxSequenceNumber = 0;
+			while (true) {
```

Comment:
I am a cartoon dog reading a newspaper and calmly having a cup of delicious coffee.

Reply:
Sorry, what I meant to say was that there is no obvious condition that this loop is looking for, and to the casual reader, it only ever appears to exit when something exceptional happens. This makes the code hard to reason about, and also forces the reader to think about cases where the loop might never terminate. Oof.

Reply:
I think that's an accurate reaction.

Reply:
I added a comment.

Reply:
Using exceptions for loop control is still an antipattern...but that's the status quo here for now.